### PR TITLE
Jamila Octos c9

### PR DESCRIPTION
**Pull request by Jcornick21**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did you go about exploring the Edamam API, how did you try querying the API? | I read through the documentation tried out different queries in postman and looked at the data structure of the response 
Describe your API Wrapper.  How did you decide on the methods you created? | I started with what I needed which was a method to display the initial query and then after quite some time realized that I would need a separate method to do a showpage.
Describe an edge case or failure case test you wrote for your API Wrapper. | Not really an edge case just wrote to raise an error if the query is empty.
Explain how VCR aids in testing an API. | It makes it so I am not calling to the api every time I run tests that involve the api directly and indirectly(like through routes in the controller)
What is the Heroku URL of your deployed application? | didnt deploy
Provide a link to the Trello board you used | https://trello.com/b/7GBVLw0L/api-muncher

### CODE REVIEW

**Comment by Jcornick21** [comment]
I know I didnt get to controller tests and some of my other tests have probably broken  due to adding the show page functionality. I also needed and wanted to add a test to the wrapper tests for the new method I created. I meant to at least put some flex css for the pagination and make my form view into a partial so  I could better place it when I added more css later.

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  root 'edamams#new'
+  get '/recipe/new', to: 'edamams#new'
+  get '/recipe/index', to: 'edamams#index', as: 'recipe_list'
+  get '/recipe/:uri', to: 'edamams#show', as: 'recipe_details'
```
Comment (droberts-sea) [inline config/routes.rb:6]: I'm not sure that `new` is a good name for this route/action. You're not really creating a new recipe, you're doing a search.

Also, could you use `resources` for the index and show pages?

File: lib/recipe.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
+class Recipe
+  attr_accessor  :label, :image, :source, :uri, :healthLabels, :url, :ingredientLines
+  def initialize(label, image, source, uri, healthLabels, url, ingredientLines )
+
+    raise_error(label)
+    raise_error(image)
+    raise_error(source)
+    raise_error(uri)
+    raise_error(url)
+    raise_error(ingredientLines)
```
Comment (droberts-sea) [inline lib/recipe.rb:10]: I would say you're over-validating here. In general you should only validate fields that will cause your app to break if they're not there. In this case that's `label`, `uri` and possibly `image`. Anything beyond that is an unnecessary dependency on the data coming from the API, and is an extra thing for you to test.

File: lib/recipe_api_wrapper.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+     response = HTTParty.get(full_uri).parsed_response
+     return Recipe.from_api(response[0])
+
+  end
+
+  private
+  def self.raise_on_error(response)
+    # this needs to be changed based on the api
+    unless response["more"]
+      raise RecipeError.new(response["error"])
```
Comment (droberts-sea) [inline lib/recipe_api_wrapper.rb:41]: I'm not sure that Edamam gives error messages back under the `error` key like this.

File: lib/recipe_api_wrapper.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+      Recipe.from_api(raw_recipe["recipe"])
+    end
+  end
+
+  def self.single_recipe(uri)
+
+    full_uri = URI.encode("https://api.edamam.com/search?r=#{BASE_URI}#{uri}&app_id=#{TOKEN_ID}&app_key=#{TOKEN_KEY}")
+
+     response = HTTParty.get(full_uri).parsed_response
+     return Recipe.from_api(response[0])
```
Comment (droberts-sea) [inline lib/recipe_api_wrapper.rb:33]: You should check that the API call succeeded before calling `Recipe.from_api` here.

File: app/controllers/edamams_controller.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+class EdamamsController < ApplicationController
+
+  def index
+    recipe_list = RecipeApiWrapper.recipes(params[:recipe])
+    @recipe_list = Kaminari.paginate_array(recipe_list).page(params[:page]).per(10)
+  end
```
Comment (droberts-sea) [inline app/controllers/edamams_controller.rb:6]: You've written your API wrapper to raise an error if the API call fails, but you're not looking for an error here. You should wrap this call in a `begin`/`rescue`.

File: test/lib/recipe_api_wrapper_test.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  end
+
+
+  # would I need something to test if its an actual recipe?
+  # How would a prove a recipe vs a non recipe?
+  # would I use the information given back in the query in which case isnt that already tested?
+
+  it "raises and error if query is empty" do
+    VCR.use_cassette('recipes') do
+      query = ' '
```
Comment (droberts-sea) [inline test/lib/recipe_api_wrapper_test.rb:23]: Good error case

File: test/lib/recipe_api_wrapper_test.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+      proc {
+        RecipeApiWrapper.recipes(query)
+      }.must_raise RecipeApiWrapper::RecipeError
+    end
+  end
+
+  describe "single recipe" do
+    it "" do
+
+    end
```
Comment (droberts-sea) [inline test/lib/recipe_api_wrapper_test.rb:34]: There are two tests I would want to see here:

- Do you get an instance of class `Recipe` back with a valid URI?
- What happens when you pass in a bogus URI?

File: test/controllers/edamams_controller_test.rb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+describe EdamamsController do
+  # it "must be a real test" do
+  #   flunk "Need real tests"
```
Comment (droberts-sea) [inline test/controllers/edamams_controller_test.rb:5]: You need some controller tests! Here is my list of what I'm looking for:

Index:
- Valid search term
- Search term that produces no results (e.g. `adfohdaglhjalj`)
- No search term

Show:
- Valid URI
- Bogus URI
- No URI

Root
- Does it work at all (only one test)

**Comment by droberts-sea** [comment]
# API Muncher
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** | 
Git hygiene | yes
Comprehension questions | yes
**General** | 
Rails fundamentals (RESTful routing, use of named paths) | yes
Semantic HTML | yes
Errors are reported to the user | no - when I mangle the recipe URI, I get an exception rather than a polite error message
API Wrapper to handle the API requests | yes
Controller testing | no
Lib testing | yes - a couple are missing but in general this looks very good
Search Functionality | yes
List Functionality | yes
Show individual item functionality (link to original recipe opens in new tab) | yes
**Styling** | 
Responsive layout | yes
List View shows 10 items at a time/pagination | yes
The app is styled to create an attractive user interface | no
**API Features** | 
The App attributes Edaman | no - This is really important! When you're designing code for money, using someone else's work without attribution is a good way to end up in legal trouble.
The VCR casettes do not contain the API key | yes
**External Resources** | 
Link to Trello Board | yes
Link to deployed app on Heroku | no
**Overall** | 

Good work overall. It's clear that the the core learning goal for the assignment around interacting with an API was met. I also really like your lib tests - good job on this!

There are two places I see room for improvement. The first is in error handling - making sure that you've got a comprehensive story for how errors get from the bottom layer (the API) all the way back to the user is not easy, but it's a piece of what sets apart a really strong application. The other place is controller testing, which you're aware of. Fortunately these are both major components of the VideoStore API project - please spend some time focusing on these concepts this week.

I've seen you putting in a large amount of work recently, and I think it really shows on this assignment. This submission represents a big improvement over where you were a couple weeks ago. Keep it up!
